lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -2646,9 +2646,10 @@ bool TypeChecker::typeCheckBinding(Pattern *&pattern, Expr *&initializer,
   auto contextualPurpose = CTP_Unused;
   TypeCheckExprOptions flags = TypeCheckExprFlags::ConvertTypeIsOnlyAHint;
 
+  contextualPurpose = CTP_Initialization;
```

Comment:
Would be great to leave a comment here in order to clarify why this is a good idea.